vcstool/commands/compare.py

```diff
@@ -1,80 +1,595 @@
+
+    EQUAL = 0
+    LOCAL = 1
+    LOCAL_BEHIND_REMOTE = 2
+    LOCAL_AHEAD_OF_REMOTE = 3
+    DIVERGED = 4
+    ERROR = 5
+
+
+class ICompareTableEntry(abc.ABC):
+    """Interface used to implement an entry (row) in the CompareTable."""
+
+    HEADERS = ["S", "Repository", "Branch", "Trk", "Flags", "Tag", "Hash"]
+
+    def get_color_row(self, is_odd_row: bool, num_cols: int) -> List[str]:
+        """Returns a formatted and colored row representing this entry."""
+        # The order of these entries should match the order of HEADERS.
+        row = [
+            self.get_color_repo_status(),
+            self.get_color_repository(),
+            self.get_color_branch(),
+            self.get_color_track(),
+            self.get_color_vcs_tracking_flags(),
+            self.get_color_tag(),
+            self.get_color_hash(),
+        ]
+        row = self._wrap_row_with_background_color(row, is_odd_row)
+        # max_dispaly_cols is used to hide columns if the terminal is not wide enough to display
+        # the full table.
+        return row[:num_cols]
+
+    @staticmethod
+    def _wrap_row_with_background_color(row: List[str], is_odd_row: bool):
+        reset = ansi("reset")
+        background = ansi("grey4b") if is_odd_row else reset
+        return [background + item + reset + background for item in row]
+
+    @abc.abstractmethod
+    def is_significant(self) -> bool:
+        return NotImplemented
+
+    @abc.abstractmethod
+    def legend_flags(self) -> Legend.Flags:
+        return NotImplemented
+
+    @abc.abstractmethod
+    def get_color_repo_status(self) -> str:
+        return NotImplemented
+
+    @abc.abstractmethod
+    def get_color_repository(self) -> str:
+        return NotImplemented
+
+    @abc.abstractmethod
+    def get_color_branch(self) -> str:
+        return NotImplemented
+
+    @abc.abstractmethod
+    def get_color_track(self) -> str:
+        return NotImplemented
+
+    @abc.abstractmethod
+    def get_color_vcs_tracking_flags(self) -> str:
+        return NotImplemented
+
+    @abc.abstractmethod
+    def get_color_tag(self) -> str:
+        return NotImplemented
+
+    @abc.abstractmethod
+    def get_color_hash(self) -> str:
+        return NotImplemented
+
+
+class CompareTable(pt.PrettyTable):
+    """PrettyTable extension which displays a table describing the state of the workspace."""
+
+    def __init__(
+        self,
+        entries: Dict[str, ICompareTableEntry],
+        manifest_file: str,
+        max_width: int = os.get_terminal_size().columns,
+    ) -> None:
+        super().__init__()
+
+        self._entries = entries
+        self._manifest_file = manifest_file
+        self._legend_flags = Legend.Flags(0)
+
+        # Initial generation.
+        num_cols = len(ICompareTableEntry.HEADERS)
+        self._reset_and_add_entries(num_cols)
+
+        # If the table width is too wide, continually remove columns from the right.
+        # TODO(amusco): More efficient way to do this?
```

Comment:
Nit: should move this *TODO* to a ticket and reference here, or accept efficiency and remove. Again, not a python master so I dont have a good alternative. :)

Reply:
Yep totally agree! I typically add TODO's in PRs to take notes, but once it gets closer to merging I'll take them out and cut tickets. Not a fan of leaving TODOs in production code (code smell).

Reply:
Agree with removing TODOs for production code and moving to tickets.